validator/client/wait_for_activation.go

```diff
@@ -13,22 +12,25 @@ import (
 	"github.com/prysmaticlabs/prysm/shared/params"
 	"github.com/prysmaticlabs/prysm/shared/slotutil"
 	"github.com/prysmaticlabs/prysm/shared/traceutil"
-	"github.com/sirupsen/logrus"
 	"go.opencensus.io/trace"
 )
 
 // WaitForActivation checks whether the validator pubkey is in the active
 // validator set. If not, this operation will block until an activation message is
 // received. This method also monitors the keymanager for updates while waiting for an activation
 // from the gRPC server.
-func (v *validator) WaitForActivation(ctx context.Context) error {
+//
+// If the channel parameter is nil, WaitForActivation creates and manages its own channel.
+func (v *validator) WaitForActivation(ctx context.Context, accountsChangedChan chan [][48]byte) error {
 	// Monitor the key manager for updates.
-	accountsChangedChan := make(chan [][48]byte)
-	sub := v.GetKeymanager().SubscribeAccountChanges(accountsChangedChan)
-	defer func() {
-		sub.Unsubscribe()
-		close(accountsChangedChan)
-	}()
+	if accountsChangedChan == nil {
+		accountsChangedChan = make(chan [][48]byte)
+		sub := v.GetKeymanager().SubscribeAccountChanges(accountsChangedChan)
+		defer func() {
+			sub.Unsubscribe()
+			close(accountsChangedChan)
```

Comment:
This defer will close the channel by the time it makes it into v.waitForActivation. Is this intended behavior?

Reply:
The function will not return until `waitForActivation` completes, so the channel will not get closed. This PR only adds an `if` statement, so it wouldn't have worked in the first place.

Reply:
You are totally right